Cleanup of consolewininfo.cpp

npwoods opened the pull request:
- Moved private static methods to an anonymous namespace
- Made better usage of string classes
- Moved invocations of Get[Open|Save]FileName() to function, to reduce copy and paste code
- Saving and restoring working directory around calls to Get[Open|Save]FileName(), to not break relative paths in settings (this fixes a problem Robert found)

This class is a mess.  The functionality of using GetOpenFileName() to try to open a software list makes some very eggregious assumptions, but this change is still an improvement.

cuavas commented on src/osd/modules/debugger/win/consolewininfo.cpp:
```diff
@@ -20,10 +20,185 @@
-consolewin_info::consolewin_info(debugger_windows_interface &debugger) :
-	disasmbasewin_info(debugger, true, "Debug", nullptr),
-	m_devices_menu(nullptr)
+using namespace std::literals;
```
No using namespace at file scope - it's too dangerous.

cuavas commented on src/osd/modules/debugger/win/consolewininfo.cpp:
```diff
@@ -20,10 +20,185 @@
+//  copy_extension_list
+//-------------------------------------------------
+
+void copy_extension_list(std::ostringstream &dest, const char *extensions)
```
Why does this care that it's an `ostringstream` in particular?

cuavas commented on src/osd/modules/debugger/win/consolewininfo.cpp:
```diff
@@ -20,10 +20,185 @@
+//  add_filter_entry
+//-------------------------------------------------
+
+void add_filter_entry(std::ostringstream &dest, const char *description, const char *extensions)
```
Same as above, shouldn't it be happy with any old `ostream`?

cuavas commented on src/osd/modules/debugger/win/consolewininfo.cpp:
```diff
@@ -20,10 +20,185 @@
+	selected_filename[0] = '\0';
+
+	// record the current working directory
+	DWORD working_directory_length = GetCurrentDirectory(0, nullptr);
```
This isn't thread-safe since the working directory is per-process.  The web server and lua engine are on work threads, so they'll still get screwed up by this.

cuavas commented on src/osd/modules/debugger/win/consolewininfo.cpp:
```diff
@@ -90,11 +265,19 @@ consolewin_info::consolewin_info(debugger_windows_interface &debugger) :
 
 
+//-------------------------------------------------
+//  dtor
```
Please don't do this.  This comment is pure noise.  I can see what a destructor is.

cuavas commented on src/osd/modules/debugger/win/consolewininfo.cpp:
```diff
@@ -90,11 +265,19 @@ consolewin_info::consolewin_info(debugger_windows_interface &debugger) :
 
 
+//-------------------------------------------------
+//  set_cpu
```
This is another comment that adds no value, just clutter, additional work to update, and risk of getting out of sync with the actual code.

cuavas commented on src/osd/modules/debugger/win/consolewininfo.cpp:
```diff
@@ -112,6 +295,10 @@ void consolewin_info::set_cpu(device_t &device)
 
 
+//-------------------------------------------------
+//  recompute_children
```
Same with this, and the rest of them.  This increases line count for no useful purpose.

cuavas commented:
Windows standard file dialogs are nasty.  You can't pass in a desired "starting directory", even if you pass `OFN_NOCHANGEDIR` it apparently still changes the working directory while browsing and just restores it afterwards, and that flag only works for save, not open.  It seems you can't safely use them in a multi-threaded application, hence people writing replacements.

npwoods commented:
I've addressed concerns on this thread, made further cleanups, and merged from master.

It isn't clear to me what to do about the current working directory nonsense.  `OFN_NOCHANGEDIR` only works for saving; while I could change the logic to use `OFN_NOCHANGEDIR` for saving and keep it for opening, is that really a good idea?

rb6502 commented:
I'd instate a rule that things on work threads should just getcwd() or OSD equivalent at the start of MAME, allowing the foreground emulation to be free to mess with the real OS working directory.

npwoods commented:
@rb6502 i'm open to that possibility; i'll let @cuavas be "the decider" on that one

npwoods commented:
if we take this approach, i'll create a separate PR to take that approach

cuavas commented:
I still want you to remove all the zero-value comments.  They're pointless clutter.
